engines/freescape/gfx.cpp

```diff
@@ -0,0 +1,154 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/config-manager.h"
+#include "graphics/renderer.h"
+#include "graphics/surface.h"
+#if defined(USE_OPENGL_GAME) || defined(USE_OPENGL_SHADERS) || defined(USE_GLES2)
+#include "graphics/opengl/context.h"
+#endif
+
+#include "engines/util.h"
+#include "math/glmath.h"
+
+#include "engines/freescape/gfx.h"
+
+namespace Freescape {
+
+Renderer::Renderer(OSystem *system, int screenW, int screenH)
+	: _system(system) {
+
+	_screenW = screenW;
+	_screenH = screenH;
+	_currentPixelFormat = Graphics::PixelFormat(4, 8, 8, 8, 8, 24, 16, 8, 0);
+	_originalPixelFormat = Graphics::PixelFormat::createFormatCLUT8();
+	_palettePixelFormat = Graphics::PixelFormat(3, 8, 8, 8, 0, 0, 8, 16, 0);
+	_keyColor = -1;
+	_palette = nullptr;
+	_colorMap = nullptr;
+	_isAmiga = false;
+	_isAtariST = false;
+}
+
+Renderer::~Renderer() {}
+
+Graphics::Surface *Renderer::convertFromPalette(Graphics::PixelBuffer *rawsurf) {
+	Graphics::Surface *surf = new Graphics::Surface();
+	surf->create(_screenW, _screenH, _originalPixelFormat);
+	surf->copyRectToSurface(rawsurf->getRawBuffer(), surf->w, 0, 0, surf->w, surf->h);
+	surf->convertToInPlace(_currentPixelFormat, _palette->getRawBuffer());
+	return surf;
+}
+
+bool Renderer::getRGBAt(uint8 index, uint8 &r, uint8 &g, uint8 &b) {
+
+	if (index == _keyColor)
+		return false;
+
+	if (index == 0) {
+		_palette->getRGBAt(0, r, g, b);
+		return true;
+	}
+
+	if (_isAmiga || _isAtariST) {
+		_palette->getRGBAt(index, r, g, b);
+		return true;
+	}
+
+	// assert(index-1 < _colorMap->size());
+	byte *entry = (*_colorMap)[index - 1];
+	uint8 color = 0;
+	uint8 acc = 1;
+	for (int i = 0; i < 4; i++) {
+		byte be = *entry;
+		if (be != 0 && be != 0xff) {
+			// TODO: fix colors for non-DOS releases
+			_palette->getRGBAt(index, r, g, b);
+			return true;
+		}
+
+		if (be == 0xff)
+			color = color + acc;
+
+		acc = acc << 1;
+		entry++;
+	}
+	assert(color < 16);
+	_palette->getRGBAt(color, r, g, b);
+	return true;
+}
+
+Common::Rect Renderer::viewport() const {
+	return _screenViewport;
+}
+
+void Renderer::computeScreenViewport() {
+	_screenViewport = Common::Rect(_screenW, _screenH);
+}
+
+Renderer *createRenderer(OSystem *system, int screenW, int screenH) {
+	Common::String rendererConfig = ConfMan.get("renderer");
+	Graphics::PixelFormat pixelFormat = Graphics::PixelFormat(4, 8, 8, 8, 8, 24, 16, 8, 0);
+	Graphics::RendererType desiredRendererType = Graphics::kRendererTypeTinyGL;  // Graphics::parseRendererTypeCode(rendererConfig);
```

Comment:
This is hardcoded until we have more than one renderer (and a good way to select which one to use). I simplified a little the code to remove this code, but it will be expanded as soon as we got the OpenGL renderer.